modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/IncrementalSnapshotVerificationTask.java

```diff
@@ -0,0 +1,421 @@
+        /** Checks that current baseline topology matches baseline topology of the snapshot. */
+        private void checkBaseline(BaselineTopology blt) throws IgniteCheckedException, IOException {
+            IgniteSnapshotManager snpMgr = ignite.context().cache().context().snapshotMgr();
+
+            File snpDir = snpMgr.snapshotLocalDir(snpName, snpPath);
+            SnapshotMetadata meta = snpMgr.readSnapshotMetadata(snpDir, ignite.localNode().consistentId().toString());
+
+            String errMsg = "Topologies of snapshot and current cluster are different [snp=" +
+                meta.baselineNodes() + ", current=" + blt.consistentIds() + ']';
+
+            if (blt.size() != meta.baselineNodes().size())
+                throw new IgniteCheckedException(errMsg);
+
+            for (Object consistentId: blt.consistentIds()) {
+                if (!meta.baselineNodes().contains(consistentId.toString()))
+                    throw new IgniteCheckedException(errMsg);
+            }
+        }
+
+        /** @return Collection of snapshotted transactional caches, key is a cache ID. */
+        private Map<Integer, StoredCacheData> readTxCachesData() throws IgniteCheckedException, IOException {
+            File snpDir = ignite.context().cache().context().snapshotMgr().snapshotLocalDir(snpName, snpPath);
+
+            String folderName = ignite.context().pdsFolderResolver().resolveFolders().folderName();
+
+            File snpDbDir = new File(snpDir, databaseRelativePath(folderName));
+
+            Map<Integer, StoredCacheData> cacheData = new HashMap<>();
+
+            File[] caches = snpDbDir.listFiles();
+
+            for (File cache: caches) {
+                File[] cfgs = cache.listFiles(p -> Files.isRegularFile(p.toPath()) && p.getName().endsWith(CACHE_DATA_FILENAME));
+
+                for (File cfg: cfgs) {
+                    StoredCacheData data = ignite.context().cache().configManager().readCacheData(cfg);
+
+                    if (data.config().getAtomicityMode() == CacheAtomicityMode.TRANSACTIONAL)
+                        cacheData.put(data.cacheId(), data);
+                }
+            }
+
+            return cacheData;
+        }
+
+        /** {@inheritDoc} */
+        @Override public boolean equals(Object o) {
+            if (this == o)
+                return true;
+
+            if (o == null || getClass() != o.getClass())
+                return false;
+
+            VerifyIncrementalSnapshotJob job = (VerifyIncrementalSnapshotJob)o;
+
+            return snpName.equals(job.snpName) && Objects.equals(incIdx, job.incIdx) && Objects.equals(snpPath, job.snpPath)
+                && consId.equals(job.consId);
+        }
+
+        /** {@inheritDoc} */
+        @Override public int hashCode() {
+            return Objects.hash(snpName, incIdx, snpPath, consId);
+        }
+    }
+
+    /** Holder for parition hashes. */
+    private static class PartitionHashHolder {
+        /** */
+        private int hash;
```

Comment:
Looks like these fields should be long

Reply:
I made it as `int`, because `PartitionHashRecordV2` stores hashes as `int` too: `partHash`, `partVerHash`.